apps/src/blockly/addons/cdoBlockSvg.js

```diff
@@ -70,20 +70,52 @@ export default class BlockSvg extends GoogleBlockly.BlockSvg {
           Blockly.ContextMenu.hide();
         }.bind(this)
       };
-      const lockToParent = {
-        text: this.canDisconnectFromParent_
-          ? 'Lock to Parent Block'
-          : 'Unlock from Parent Block',
-        enabled: true,
-        callback: function() {
-          this.setCanDisconnectFromParent(!this.canDisconnectFromParent_);
-          Blockly.ContextMenu.hide();
-        }.bind(this)
-      };
       menuOptions.push(deletable);
       menuOptions.push(movable);
       menuOptions.push(editable);
-      menuOptions.push(lockToParent);
+      // menuOptions for Shadow Blocks require awareness of parents and children.
+      const children = this.getChildren();
+      const shadowChildCount = children.filter(child => child.isShadow())
+        .length;
+      const nonShadowChildrenCount = children.filter(child => !child.isShadow())
+        .length;
+      // A block can be made into a shadow if:
+      //* It has a surrounding parent block.
+      //* It does not contain a variable field.
+      //* It does not have any non-shadow children.
+      if (
+        this.getSurroundParent() &&
+        !this.getVarModels().length &&
+        !nonShadowChildrenCount
+      ) {
+        const shadow = {
+          text: 'Make Shadow',
+          enabled: true,
+          callback: function() {
+            this.setShadow(true);
+            Blockly.ContextMenu.hide();
+          }.bind(this)
+        };
+        menuOptions.push(shadow);
+      }
+      // If a block has shadow child(ren), it can be used to convert them to blocks.
+      if (children.length) {
+        if (shadowChildCount) {
+          const unshadow = {
+            text: `Make ${
+              shadowChildCount > 1 ? `${shadowChildCount} ` : ''
+            }Child Block${shadowChildCount > 1 ? 's' : ''} Non-Shadow`,
+            enabled: true,
+            callback: function() {
+              for (let i = 0; i < children.length; i++) {
+                children[i].setShadow(false);
+              }
+              Blockly.ContextMenu.hide();
+            }.bind(this)
```

Comment:
similarly here, there could be a method called `hasShadowChildren()` that returns a boolean based on if `children.length && shadowChildCount > 0`  or just using that expression to simplify the `if` statement.